### PR TITLE
added state root to txBlockHeader

### PR DESCRIPTION
**Pull request by sidutta**

### CODE REVIEW

File: build_lookup.sh [hunk 1]
```diff
@@ -19,5 +19,9 @@ rm -rf ./txblocks.db
 rm -rf ./test.db
 rm -rf ./txbodies.db
 
+find . -name *.cmake | xargs rm -rf
```
Comment (sharwell) [inline build_lookup.sh]: 💡 Another option is to not have this in the script, but instead run the following to clean *everything* in the working directory:

```
git clean -dxf
```

💡 It's good to make changes like this in their own commits, since it's not directly related to the rest of this pull request.

File: src/libData/BlockData/Block/TxBlock.cpp [hunk 2]
```diff
@@ -28,7 +28,8 @@ unsigned int TxBlock::Serialize(vector<unsigned char> & dst, unsigned int offset
 
     unsigned int header_size_needed = sizeof(uint8_t) + sizeof(uint32_t) + UINT256_SIZE + UINT256_SIZE + 
                                       BLOCK_HASH_SIZE + UINT256_SIZE + UINT256_SIZE + TRAN_HASH_SIZE + 
-                                      sizeof(uint32_t) + sizeof(uint32_t) + PUB_KEY_SIZE + UINT256_SIZE + BLOCK_HASH_SIZE;
+                                      TRAN_HASH_SIZE + sizeof(uint32_t) + sizeof(uint32_t) + PUB_KEY_SIZE + 
+                                      UINT256_SIZE + BLOCK_HASH_SIZE;
```
Comment (sharwell) [inline src/libData/BlockData/Block/TxBlock.cpp]: :bulb: This expression appears several times. Consider extracting a constant.

File: src/libData/BlockData/BlockHeader/TxBlockHeader.cpp [hunk 3]
```diff
@@ -169,6 +174,11 @@ const TxnHash & TxBlockHeader::GetTxRootHash() const
     return m_txRootHash;
 }
 
+const StateHash & TxBlockHeader::GetStateRootHash() const
```
Comment (sharwell) [inline src/libData/BlockData/BlockHeader/TxBlockHeader.cpp]: :bulb: Consider defining this method where it is declared (improves efficiency with no apparent downside).

File: src/libNode/FinalBlockProcessing.cpp [hunk 4]
```diff
@@ -888,7 +888,7 @@ bool Node::ProcessFinalBlock(const vector<unsigned char> & message, unsigned int
     if(m_state == MICROBLOCK_CONSENSUS)
     {
         unsigned int time_pass = 0;
-        while(m_state != PROCESS_FINALBLOCK)
+        while(m_state != WAITING_FINALBLOCK)
```
Comment (sharwell) [inline src/libNode/FinalBlockProcessing.cpp]: ❔ Why did this change? It seems unrelated to the rest of this PR.

**Comment by sharwell** [comment]
@sidutta Can you briefly describe the relation between this pull request and #32?

File: build_lookup.sh [hunk 5]
```diff
@@ -19,5 +19,9 @@ rm -rf ./txblocks.db
...
+find . -name CMakeLists.txt | xargs rm -rf
```
Comment (sharwell) [inline build_lookup.sh]: ❗️ This line breaks the build by deleting the CMake input.

File: src/libData/BlockData/BlockHeader/TxBlockHeader.cpp (hunk 3 above)
Comment (sidutta) [reply src/libData/BlockData/BlockHeader/TxBlockHeader.cpp]: Sure.

File: src/libNode/FinalBlockProcessing.cpp (hunk 4 above)
Comment (sidutta) [reply src/libNode/FinalBlockProcessing.cpp]: This was a bug that had to be corrected.

**Comment by sidutta** [comment]
@sharwell The commits 32d80a0, e3035c8, b390ebd nearly complete #32. The last mini epoch of every major epoch has been made vacuous(the microblock in that epoch doesn't have any transactions). This should give (hopefully) sufficient time for the DS nodes to receive transaction bodies from all the prior mini epochs. In the commit 32d80a0 I added a new field in the final block header to store the state root. In e3035c8, I added some synchronization barriers so that DS node leader has all transaction bodies before it attempts to create the final block. In b390ebd, I compute the state root and store it in the field created as part of 32d80a0.

File: build_lookup.sh (hunk 5 above)
Comment (sidutta) [reply build_lookup.sh]: I corrected it. It should have been CMakeCache. Thanks.

File: src/libData/BlockData/Block/TxBlock.cpp (hunk 2 above)
Comment (sidutta) [reply src/libData/BlockData/Block/TxBlock.cpp]: Yeah, even I was wondering the same when making the change.

File: src/libData/BlockData/BlockHeader/TxBlockHeader.cpp (hunk 3 above)
Comment (sidutta) [reply src/libData/BlockData/BlockHeader/TxBlockHeader.cpp]: For such changes though could you rather make generalized github issues. May be as we grow someone from the community can pick them up. We could tag such enhancements something so that they can be found.
